Add Ubuntu nodes on Vagrant env

ycointe opened the pull request:
**Component**: Vagrant

<!-- E.g. 'salt', 'containers', 'kubernetes', 'build', 'tests'... -->

**Context**: Metalk8s over Ubuntu

**Summary**: Permits to have some Ubuntu nodes on Vagrant with python 2.7 installed

**Acceptance criteria**: 
Deploy a node (with a node number > 5 ) and verify if it's a Ubuntu VM. 
Then launch the command `python` and verify the version (python 2.7.15+)


---

<!-- Declare one or more issues to close once this PR gets merged -->

Closes: #1327 #1326 

<!-- If you want to refer to an issue while not closing it, use:

See: #ISSUE_NUMBER

-->

bert-e commented:
# Hello ycointe,



My role is to assist you with the merge of this
pull request. Please type `@bert-e help` to get information
on this process, or consult the [user documentation](
https://eve.devsca.com/github/scality/metalk8s/bert-e/doc/user).





*Status report is not available.*

bert-e commented:
# Integration data created



I have created the integration data for the additional destination branches.

* this pull request will merge `feature/ubuntk8s` into
`development/2.0`
* [w/2.1/feature/ubuntk8s](https://github.com/scality/metalk8s/compare/development/2.1...w/2.1/feature/ubuntk8s) will be merged into `development/2.1` 
* [w/2.2/feature/ubuntk8s](https://github.com/scality/metalk8s/compare/development/2.2...w/2.2/feature/ubuntk8s) will be merged into `development/2.2` 



The following branches will **NOT** be impacted:

* `development/1.0`
* `development/1.1`
* `development/1.2`
* `development/1.3`




You can set option `create_pull_requests` if you need me to create
**integration pull requests** in addition to integration branches, with:

```
@bert-e create_pull_requests
```

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @jbertran

* @nootal

* @cmonfort

* @lucieleonard

* @alxf

* @carlito-scality

* @alexandre-allard-scality

* @teddyandrieux

* @slaperche-scality

* @thomasdanan

* @zempashi

* @nicolast

* @sayf-eddine-scality

* @monpote

* @gdemonet

* @tcarmet

gdemonet reviewed (dismissed):
Renaming the `salt/metalk8s/repo/offline.sls` cannot be done without the related changes in the buildchain

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @jbertran

* @nootal

* @cmonfort

* @lucieleonard

* @alxf

* @carlito-scality

* @alexandre-allard-scality

* @teddyandrieux

* @slaperche-scality

* @thomasdanan

* @zempashi

* @nicolast

* @sayf-eddine-scality

* @monpote

* @gdemonet

* @tcarmet





The following reviewers are expecting changes from the author, or must review again:

* @gdemonet

bert-e commented:
# History mismatch



Merge commit #e181d1a3eef99c9bca6d07508e0470c20fe7e6a4 on the integration branch
`w/2.1/feature/ubuntk8s` is merging a branch which is neither the current
branch `feature/ubuntk8s` nor the development branch
`development/2.1`.

It is likely due to a rebase of the branch `feature/ubuntk8s` and the
merge is not possible until all related `w/*` branches are deleted or updated.

**Please use the `reset` command to have me reinitialize these branches.**

slaperche-scality commented on Vagrantfile:
```diff
@@ -152,6 +152,15 @@ fi
 SCRIPT
 
 
+INSTALL_PYTHON_ON_UBUNTU = <<-SCRIPT
```
Instead of relying of the node number (which is hackish :stuck_out_tongue: ), we could check the distribution (by grepping in `/etc/*-release`?) and install Python only if it match Ubuntu?

gdemonet replied:
Do you mean we should provision this script for every VM? I think since we hold control over the desired OS in this same file, it's fine to keep the script straightforward and Ubuntu-specific.

slaperche-scality replied:
Then we should find a cleaner way than `> 5`

At least defined a named constant instead of using `5` :stuck_out_tongue:

ycointe replied:
Change have been done for the Ubuntu nodes deployment (with python)

slaperche-scality commented on salt/metalk8s/repo/Ubuntu.sls:
```diff
@@ -0,0 +1,30 @@
+    - name: python-apt 
+
+
+Configure Kubernetes repository:
```
Why do we skip a random number of lines?

- 2 lines between `Install python-apt` and `Configure Kubernetes repository`
- 3 lines between `Configure Kubernetes repository` and `Configure Salt repository`
- 1 line between `Configure Salt repository` and `Repositories configured`.

ycointe replied:
Empty lines deleted.

slaperche-scality commented on Vagrantfile:
```diff
@@ -152,6 +152,15 @@ fi
+
+apt install python -y
+
+SCRIPT
```
You should write a proper commit message that explain why we need to install Python 2.7 manually (and also python-apt).

ycointe replied:
Explanation done in the last commit (requirement for the salt-minion)

slaperche-scality reviewed (changes requested):
IMHO, this PR could be split into 3 commits:
- the gitignore part (commit msg: explain what are those log file we want to ignore/where they come from)
- the update of the vagrantfile (commit msg: explain why we need to install manually Python 2.7)
- the new salt file for Ubuntu (commit msg: maybe explain why we need python-apt, seems not obvious to me)

and if you do want rename `offline` to `centos`, could be in a dedicated commit as well, for tracking purpose (and you can explain why the rename in the commit message (e.g.: because brace yourself Ubuntu is coming))

gdemonet commented on Vagrantfile:
```diff
@@ -199,13 +209,23 @@ Vagrant.configure("2") do |config|
     node_name = "node#{i}"
 
     config.vm.define node_name, autostart: false do |node|
+      if i > 5
```
A simple suggestion to make things more obvious for users (as @slaperche-scality mentioned, using a node's number to choose which OS to install is kind of hackish):

Maybe you could instead duplicate the block defining `node#{i}` VMs and rename them `centos#{i}` and `ubuntu#{i}`. I'm also checking around Vagrant docs to see if we could parametrize this more easily (cc @nootal, any idea?).

gdemonet replied:
Let's go for this simple solution right now, maybe we'll want to handle user configuration (through a file) in a later iteration.

NicolasT replied:
Yes, please use the OS name in the node name, and always count from 1.

ycointe replied:
Change done.

TeddyAndrieux commented on salt/metalk8s/salt/minion/installed.sls:
```diff
@@ -4,6 +4,15 @@ include :
+    - require:
+      - test: Repositories configured
+
+{%- endif %}
```
IMO you sould do the change in the macro `pkg_installed` directly as this change will be needed for all packages I think

ycointe replied:
I change it in the macro but don't remove this part used in my first tests.
I will change it.

TeddyAndrieux commented on salt/metalk8s/salt/minion/installed.sls:
```diff
@@ -4,6 +4,15 @@ include :
   - metalk8s.repo
 
 Install salt-minion:
+{%- if grains['os'] == 'CentOS' %}
```
Not filter on `os` but on `osfamily` please, as we will, likely, support RedHat (maybe, one day)

ycointe replied:
Change done.

TeddyAndrieux commented on salt/metalk8s/repo/configured.sls:
```diff
@@ -1,2 +1,2 @@
 include:
-  - .offline
+  - .{{ grains['os'] }}
```
Here I think you should also (like below) filter on `osfamily` I think if we support RedHat we will have almost the same states for both.

ycointe replied:
Change done.

NicolasT commented on buildchain/buildchain/salt_tree.py:
```diff
@@ -335,7 +335,8 @@ def _run(self) -> None:
     Path('salt/metalk8s/repo/installed.sls'),
     Path('salt/metalk8s/repo/macro.sls'),
-    Path('salt/metalk8s/repo/offline.sls'),
+    Path('salt/metalk8s/repo/CentOS.sls'),
```
Filenames are all-lowercase.

ycointe replied:
Change done. (now named rhel.sls for the old CentOS.sls and debien.sls for old Ubuntu.sls)

NicolasT commented on salt/metalk8s/macro.sls:
```diff
@@ -3,11 +3,13 @@
-    - fromrepo: {{ repo.repositories.keys() | join(',') }}
-    {%- if package.version | default(None) %}
+    - name: {{ package.name | default(name) }}
+  {%- if package.version | default(None) %}
```
Why this indentation change?

TeddyAndrieux replied:
IMHO this indentation looks better compare to the previous one (same indentation as the `set` above and only one more indentation than the macro) 

Personnaly I indent jinja action (`{% stuff %}`) independently, I mean indentation not related to the "non jinja" content

ycointe replied:
The indentation is indeed based on jinja and not on yaml.

NicolasT commented on salt/metalk8s/macro.sls:
```diff
@@ -3,11 +3,13 @@
     - update_holds: True
-    {%- endif %}
+  {%- endif %}
+  {%- if grains['os'] == 'CentOS' %}
```
What about RHEL? You'll want to dispatch on family, I guess.

ycointe replied:
change done.

NicolasT commented on salt/metalk8s/repo/Ubuntu.sls:
```diff
@@ -0,0 +1,30 @@
+  pkgrepo.managed:
+    - humanname: kubernetes
+    - name: deb http://apt.kubernetes.io/ kubernetes-xenial main
+    - file: /etc/apt/sources.list.d/kubernetes.list
```
`metalk8s.list`?

ycointe replied:
I changed it.

NicolasT commented on salt/metalk8s/repo/Ubuntu.sls:
```diff
@@ -0,0 +1,30 @@
+  pkgrepo.managed:
+    - humanname: saltstack
+    - name: deb http://repo.saltstack.com/apt/ubuntu/18.04/amd64/archive/2018.3.4 bionic main
+    - file: /etc/apt/sources.list.d/saltstack.list
```
Merge all repos we set up in the same list file (IMHO).

ycointe replied:
Do the same as what I done for /etc/apt/sources.list.d/kubernetes.list I put all the source list in metalk8s.list

NicolasT commented on salt/metalk8s/salt/minion/installed.sls:
```diff
@@ -4,6 +4,6 @@ include :
-  {{ pkg_installed('salt-minion') }}
-    - require:
-      - test: Repositories configured
+{{ pkg_installed('salt-minion') }}
```
This now looks incorrect, given unindent?

ycointe replied:
Correction done.

jbertran commented on .gitignore:
```diff
@@ -10,3 +10,5 @@ buildchain/.venv/
 
 *.pyc
+*.log
+
```
Stray newline

ycointe replied:
I corrected it

jbertran commented on Vagrantfile:
```diff
@@ -195,8 +195,13 @@ Vagrant.configure("2") do |config|
 
-  (1..9).each do |i|
-    node_name = "node#{i}"
+  INSTALL_PYTHON = <<-SCRIPT
```
Let's put this nearer to the relevant loop (i.e. just before the `ubuntu` loop).

___

Alternatively, we could have nested loops roughly looking like this (untested):

```rb
os_data = {
  "ubuntu" => {
    "name" => "ubuntu/bionic64",
    "version" => "20190514.0.0"
  },
  "centos" => {
    "name" => "centos/7",
    "version" => "1811.02"
  }
}

os_data.each do |os, os_data|
  (1..5).each do |i|
        config.vm.define node_name, autostart: false do |node|
      if os == "ubuntu"
        # Do your script
      end
      config.vm.box = os_data["name"]
      config.vm.version = os_data["version"]

      # etc etc
    end
  end
end
```
NBD if you want to keep your version, it's fine as is.

slaperche-scality replied:
+1

I prefer loop over copy-pasta, so that if one day we decide to add a settings or change an existing one (like `node.vbguest.auto_update`) we will have a single place to update instead of many.

ycointe replied:
loop now used for node deployment.

jbertran commented on Vagrantfile:
```diff
@@ -195,8 +195,13 @@ Vagrant.configure("2") do |config|
+  SCRIPT
+
+  (1..5).each do |i|
+    node_name = "centos#{i}"
```
If I look at the entire file, you're missing the box/version definitions here.

```
  node.vm.box = "centos/7"
  node.vm.box_version = "1811.02"
```

ycointe replied:
Is it a requirement ? I supposed it was defined by default  at the line 158 in the vagrant.configure

jbertran replied:
Oh, indeed. It wouldn't hurt to repeat it here (or use the loop version) instead of relying on default propagation.

ycointe replied:
I use loop, thank you for your help.

jbertran commented on salt/metalk8s/container-engine/containerd/installed.sls:
```diff
@@ -50,3 +56,4 @@ Configure registry IP in containerd conf:
           endpoint = ["http://{{ registry_ip }}:{{ registry_port }}"]
     - require:
       - pkg: Install containerd
+
```
Stray newline

ycointe replied:
I deleted it

jbertran replied:
It's still here :disappointed:

jbertran commented on Vagrantfile:
```diff
@@ -151,6 +151,7 @@ if ! grep -Fxq "$(cat .ssh/#{PRESHARED_SSH_KEY_NAME}.pub)" .ssh/authorized_keys
 fi
 SCRIPT
 
+
```
Stray newline

jbertran commented on salt/metalk8s/repo/debian.sls:
```diff
@@ -2,23 +2,22 @@ Install python-apt:
     - humanname: kubernetes
     - name: deb http://apt.kubernetes.io/ kubernetes-xenial main
-    - file: /etc/apt/sources.list.d/kubernetes.list
+    #- file: /etc/apt/sources.list.d/kubernetes.list
```
Stray comment?

ycointe replied:
Oversight, I corrected it

jbertran commented on salt/metalk8s/repo/debian.sls:
```diff
@@ -2,23 +2,22 @@ Install python-apt:
     - humanname: saltstack
     - name: deb http://repo.saltstack.com/apt/ubuntu/18.04/amd64/archive/2018.3.4 bionic main
-    - file: /etc/apt/sources.list.d/saltstack.list
+    #- file: /etc/apt/sources.list.d/saltstack.list
```
Ditto

ycointe replied:
Oversight, I corrected it

jbertran reviewed (changes requested):
Nice work.

Besides the (admittedly minor) remarks, I think we can do better with the commit history. Feel free to ping me if you want to take a look.

bert-e commented:
# Branches have diverged



This pull request's source branch `feature/ubuntk8s` has diverged from
`development/2.0` by more than 50 commits.

To avoid any integration risks, please re-synchronize them using one of the
following solutions:

* Merge `origin/development/2.0` into `feature/ubuntk8s`
* Rebase `feature/ubuntk8s` onto `origin/development/2.0`

Note: If you choose to rebase, you may have to ask me to rebuild
integration branches using the `reset` command.

TeddyAndrieux commented on salt/metalk8s/container-engine/containerd/installed.sls:
```diff
@@ -8,24 +8,30 @@
   {{ pkg_installed('container-selinux') }}
     - require:
       - test: Repositories configured
+{%- endif %}
```
Suggest you use `require_in` instead of adding extra `if` every where. 
https://docs.saltstack.com/en/latest/ref/states/requisites.html#the-in-versions-of-requisites

Something like
```
Install container-selinux:
￼  {{ pkg_installed('container-selinux') }}
￼    - require:
￼      - test: Repositories configured
    - require_in:
      - pkg: Install runc
```

For me the require in `Install containerd` for `container-selinux` is useless as `Install containerd` already require `runc` but it's not really part of your PR.

ycointe replied:
Indeed, I would like to open a new PR about the different packets required for the node installation. But I keep your suggestion in mind

jbertran commented on Vagrantfile:
```diff
@@ -195,25 +195,51 @@ Vagrant.configure("2") do |config|
+          node.vm.provision "install-python",
+            type: "shell",
+            inline: INSTALL_PYTHON        
+          node.vm.box = os_data["name"]
```
Similar to my comment about not relying on default propagation, this could be outside the `if` block :)

ycointe replied:
Indeed, I changed it

jbertran reviewed (commented):
LGTM, we'll tackle the history issues bright and early tomorrow!

slaperche-scality commented on Vagrantfile:
```diff
@@ -195,25 +195,52 @@ Vagrant.configure("2") do |config|
+      "name" => "ubuntu/bionic64",
+      "version" => "20190514.0.0"
+    }
+  }
```
very minor: in Ruby we usually:
- use symbol instead of string as key
- [use simple quote when the string don't need interpolation](https://github.com/rubocop-hq/ruby-style-guide#consistent-string-literals)

e.g.:

```ruby
  os_data = {
    centos: {
      name:    'centos/7',
      version: '1811.02'
    },
    ubuntu: {
      name:    'ubuntu/bionic64',
      version: '20190514.0.0'
    }
  }
```

and then

```diff
-        node.vm.box = os_data["name"]
-        node.vm.box_version = os_data["version"]
+        node.vm.box = os_data[:name]
+        node.vm.box_version = os_data[:version]
```

ycointe replied:
Changes done.

slaperche-scality commented on Vagrantfile:
```diff
@@ -195,25 +195,52 @@ Vagrant.configure("2") do |config|
+          type: "shell",
+          inline: DEPLOY_SSH_PUBLIC_KEY    
+      end
+    end  
```
trailing space l.244, 242, 224, 218 and 201

ycointe replied:
changes dones

slaperche-scality commented on Vagrantfile:
```diff
@@ -195,25 +195,52 @@ Vagrant.configure("2") do |config|
+
+  os_data.each do |os, os_data|
+    (1..5).each do |i|
+      node_name = os+"#{i}"
```
minor: spaces around `+`

See https://github.com/rubocop-hq/ruby-style-guide#spaces-operators

ycointe replied:
Change done

ycointe committed:
Vagrant: define Ubuntu 18LTS nodes in Vagrantfile

In preparation for supporting Ubuntu 18.04 we now
deploy 5 CentOS and Ubuntu nodes in Vagrantfile.

Closes: #1327

ycointe committed:
Vagrant: Ignore *.log Ubuntu node artifacts

The ubuntu node provisioning genrerates *.log artifacts

Issus: #1327

ycointe committed:
Vagrant: Install Python2.7 on Ubuntu nodes

Python 2.7 is a requirement for the salt-ssh bootstrap process.

Closes: #1326

bert-e commented:
# History mismatch



Merge commit #a48a4305c502549b8c032ad163cbcd14c6a73d95 on the integration branch
`w/2.1/feature/ubuntk8s` is merging a branch which is neither the current
branch `feature/ubuntk8s` nor the development branch
`development/2.1`.

It is likely due to a rebase of the branch `feature/ubuntk8s` and the
merge is not possible until all related `w/*` branches are deleted or updated.

**Please use the `reset` command to have me reinitialize these branches.**

ycointe commented:
@bert-e approve

bert-e commented:
# History mismatch



Merge commit #a48a4305c502549b8c032ad163cbcd14c6a73d95 on the integration branch
`w/2.1/feature/ubuntk8s` is merging a branch which is neither the current
branch `feature/ubuntk8s` nor the development branch
`development/2.1`.

It is likely due to a rebase of the branch `feature/ubuntk8s` and the
merge is not possible until all related `w/*` branches are deleted or updated.

**Please use the `reset` command to have me reinitialize these branches.**





*The following options are set:* **approve**

ycointe commented:
@bert-e reset

bert-e commented:
# Reset complete



I have successfully deleted this pull request's integration branches.






*The following options are set:* **approve**

bert-e commented:
# Integration data created



I have created the integration data for the additional destination branches.

* this pull request will merge `feature/ubuntk8s` into
`development/2.0`
* [w/2.1/feature/ubuntk8s](https://github.com/scality/metalk8s/compare/development/2.1...w/2.1/feature/ubuntk8s) will be merged into `development/2.1` 
* [w/2.2/feature/ubuntk8s](https://github.com/scality/metalk8s/compare/development/2.2...w/2.2/feature/ubuntk8s) will be merged into `development/2.2` 



The following branches will **NOT** be impacted:

* `development/1.0`
* `development/1.1`
* `development/1.2`
* `development/1.3`




You can set option `create_pull_requests` if you need me to create
**integration pull requests** in addition to integration branches, with:

```
@bert-e create_pull_requests
```







*The following options are set:* **approve**

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @jbertran

* @nootal

* @cmonfort

* @lucieleonard

* @alxf

* @carlito-scality

* @alexandre-allard-scality

* @teddyandrieux

* @slaperche-scality

* @thomasdanan

* @zempashi

* @nicolast

* @sayf-eddine-scality

* @monpote

* @gdemonet

* @tcarmet





The following reviewers are expecting changes from the author, or must review again:

* @nicolast

* @gdemonet







*The following options are set:* **approve**

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @jbertran

* @nootal

* @cmonfort

* @lucieleonard

* @alxf

* @carlito-scality

* @alexandre-allard-scality

* @teddyandrieux

* @slaperche-scality

* @thomasdanan

* @zempashi

* @nicolast

* @sayf-eddine-scality

* @monpote

* @gdemonet

* @tcarmet





The following reviewers are expecting changes from the author, or must review again:

* @gdemonet







*The following options are set:* **approve**

bert-e commented:
# In the queue



The changeset has received all authorizations and has been added to the
relevant queue(s). The queue(s) will be merged in the target development
branch(es) as soon as builds have passed.

The changeset will be merged in:


* :heavy_check_mark: `development/2.0`

* :heavy_check_mark: `development/2.1`

* :heavy_check_mark: `development/2.2`



The following branches will **NOT** be impacted:

* `development/1.0`
* `development/1.1`
* `development/1.2`
* `development/1.3`



There is no action required on your side. You will be notified here once
the changeset has been merged. In the unlikely event that the changeset
fails permanently on the queue, a member of the admin team will
contact you to help resolve the matter.

*IMPORTANT*

Please do not attempt to modify this pull request.

* Any commit you add on the source branch will trigger a new cycle after the
  current queue is merged.
* Any commit you add on one of the integration branches will be **lost**.

If you need this pull request to be removed from the queue, please contact a
member of the admin team now.




*The following options are set:* **approve**

bert-e commented:
# 


I have successfully merged the changeset of this pull request
into targetted development branches:


* :heavy_check_mark: `development/2.0`

* :heavy_check_mark: `development/2.1`

* :heavy_check_mark: `development/2.2`



The following branches have **NOT** changed:

* `development/1.0`
* `development/1.1`
* `development/1.2`
* `development/1.3`



Please check the status of the associated issue None.

Goodbye ycointe.
